Naming Customization and Cleanup

erikerikson opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->
## What did you implement:

**Closing:** #2022 

<!--
Briefly describe the feature if no issue exists for this PR
-->
## How did you implement it:

<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

I added `~/lib/plugins/aws/lib/naming.js` as a central definition of naming standards.   All code partaking in naming was altered to use the centralized naming capability (or at least this was the intention).  The notable exceptions to this is the not-provider-specific function name (an overridable default) and the packageService defined names of zip files.
## How can we verify it:

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* AWS CLI commands - To list AWS resources and show that the correct config is in place
* Other - Anything else that comes to mind to help us evaluate
-->

Tests were added for the new code.  That said, the change was intended to be a no-op from a correctness standpoint.  Where names were hard coded, the naming class is now used.  Otherwise, nothing of semantic importance should be changed.  I found it useful to corrupt the centrally defined names in order to identify areas that may have been missed.  The relationship between hardcoded .json files and the centralized naming is the only source of errors that I found.
## Todos:
- [x] Write tests
- [x] Write documentation (N/A)
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [] Change ready for review message below

**_Is this ready for review?:**_ NO

erikerikson commented:
This should be ready for review.

Note that I only attempted to centralize naming.  I did not yet attempt to add controls over naming conventions or anything of the like (although these seem vitally important).

I believe that some of the idioms will be unpleasant and should be fixed (particularly the `naming.configure(serverless)` idiom) but I thought it would be better to open those to discussion.

erikerikson commented:
Reduced coverage is as a result of reduced lines of code.

eahefnawy commented:
Beautiful! Thanks for all the heavy lifting @erikerikson 😊 ... our codebase was starting to get bloated because of all that naming issues.

Currently reviewing...

dougmoscrop commented on lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js:
```diff
@@ -1,7 +1,10 @@
+const BbPromise = require('bluebird');
+const pathLib = require('path');
+
+const naming = require(pathLib.join(__dirname, '..', '..', '..', '..', '..', 'lib', 'naming.js'));
```
Could you mix-in naming in to the apiGateway/index.js file to avoid dot dot hell?

erikerikson replied:
Absolutely, I thought I'd removed all instances of this from the code.  Thanks for noticing it.

erikerikson replied:
Oh...  I misread this initially.  I've removed the painful use of path.  There was a set of cases where my search had failed to catch some occurrences.

Reading you as requesting that naming.js be moved into `~/lib/plugins/aws/deploy/compile/events/apiGateway/index.js`, I have to note that the naming module is defined at `~/lib/plugins/aws/lib/naming.js` because the naming conventions need to be applied through the provider specific plugin.  Examples include:
`~/lib/plugins/aws/index.js`
`~/lib/plugins/aws/info/index.js`
`~/lib/plugins/aws/logs/index.js`
`~/lib/plugins/aws/remove/index.js`

dougmoscrop commented:
Ya I was wondering maybe putting it on provider? Like

serverless.provider.nameXYZ(x)

Make the aws plugin /lib object.assign it to provider or something.

On Tue, Oct 4, 2016, 6:56 PM Erik Erikson <notifications@github.com> wrote:

> *@erikerikson* commented on this pull request.
> ------------------------------
>
> In lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js
> <https://github.com/serverless/serverless/pull/2257>:
>
> >  const _ = require('lodash');
> +const BbPromise = require('bluebird');
> +const pathLib = require('path');
> +
> +const naming = require(pathLib.join(__dirname, '..', '..', '..', '..', '..', 'lib', 'naming.js'));
>
> Oh... I misread this initially. I've removed the painful use of path.
> There was a set of cases where my search had failed to catch some
> occurrences.
>
> Reading you as requesting that naming.js be moved into
> ~/lib/plugins/aws/deploy/compile/events/apiGateway/index.js, I have to
> note that the naming module is defined at ~/lib/plugins/aws/lib/naming.js
> because the naming conventions need to be applied through the provider
> specific plugin. Examples include:
> ~/lib/plugins/aws/index.js
> ~/lib/plugins/aws/info/index.js
> ~/lib/plugins/aws/logs/index.js
> ~/lib/plugins/aws/remove/index.js
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2257>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFAVn4qDLSKKgh_aOV-5xzzLJyTryks5qwtmzgaJpZM4KNNHf>
> .
>

erikerikson commented:
Okay, I see what you're suggesting.  I'll check but there shouldn't be any trouble doing that.

eahefnawy commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,234 @@
+  // TODO the frequency of using this method should be low so maybe it doesn't matter.
+  // TODO better suggestions highly welcome
+  configure(serverless) {
+    this.serverless = serverless;
```
why not in the constructor like we're doing in all of our plugins?

erikerikson replied:
Yes, moved it to the constructor in the latest commit in response to @dougmoscrop's great feedback.  Still absorbing idioms here.

eahefnawy commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,234 @@
+        '-',
+        [
+          this.serverless.config.stage,
+          // TODO is it a bug here in that the role is not specific to the region?
```
I don't think most AWS resource need to be specific to a region, only S3 as far as I know because they're global. But so far we haven't had an issu with that. Other examples are the Lambda name and stack name, it's not specific to the region either.

erikerikson replied:
What was in mind when I wrote that comment is that if you are doing cross-region deployments that you are likely going to need to do cross-region replication or at least backup and that you would generally want to manage that behind region-specific resources.  I suppose so long as we're only talking about default privileges that this is over-thinking the matter.

eahefnawy replied:
@erikerikson that's a super interesting user case! I haven't thought about that before! Do you think this would cause an issue for you guys at Nordstrom?

erikerikson replied:
In the short term cross-region deployments are not wide spread but we there are definitely those doing this.

eahefnawy commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,234 @@
+  }
+  extractLambdaNameFromArn(functionArn) {
+    const splitArn = functionArn.split(':');
+    // TODO the following two lines assumes default function naming?  Is there a better way?
```
I think it depends on the context, but this insures we get the actual **Lambda Name** whether the user provided a custom one via the `name` property, or the default name.

erikerikson replied:
if the user's custom name is something like `my-name-with-dashes`, wouldn't it provide the value `dashes`?  We definitely have names with components separated by dashes, which is why it occurred to me.

erikerikson replied:
Sorry, I think I wasn't quite clear.  This is from [authorizers.js#30](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/apiGateway/lib/authorizers.js#L30).  It is extracting a function name from a given ARN that should identify a lambda that is to be used as the authorizer for apig.  I don't believe that should contain the logical name that was used in the CFT to create the Lambda.  Instead, I believe that if I set a function's custom `name` to be something like `a-name-with-dashes` that I'll end up with an ARN that is something like `arn:aws:lambda:region:${accountId}:function:a-name-with-dashes`.

I would expect this code to return `dashes` rather than `a-name-with-dashes` which I believe is the expected extraction.

Another thought is that the following would seem to be more a appropriate authorizer ARN specification (just adding stage):
`arn:aws:lambda:region:${accountId}:function:a-name-with-dashes:${stage}`
In that case, the stage would be what gets split.

Sorry for being a bit confused and context-poor here, just asking questions from my curiosity while trying to make sure and capture all the name provision and manipulation logic.

eahefnawy replied:
aaaaah yeaah I think you're right 👍 ... in the LOC you provided we're assuming that splitting the dashes will **always** split into only function/stage, but if the function name has dashes the function name itself would be split too.

That's what you mean right? 😄 

does the `extractLambdaNameFromArn` method you provided fixes that?

erikerikson replied:
That is a much clearer statement of what I meant.  😄 
No, I haven't tried to change any of this logic, I am merely working to centralize it.
My thinking is that once this is centralized, we can deep dive on this type of matter as well as enabling configuration and/or plug-ability

eahefnawy commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,234 @@
+   * S3 Bucket Naming
+   */
+  getLogicalDeploymentBucketName() {
+    return 'ServerlessDeploymentBucket';
```
would that work with the recent custom bucket feature we merged few days ago? we might need another method or something and this one could be called `getLogicalDeploymentBucketName`

erikerikson replied:
Hmm...  I'm not sure.  I have a few commits that need to be merged in.  If it is already merged and there is a test that evaluates custom bucket declaration, then it does work.  If there isn't, then I will run into this when I do the merge.  I've mostly attempted to not change the consequences of the code at all while centralizing the naming so it should have the exact behavior it did before my changes.

erikerikson replied:
To explain why: basically, this "LogicalDeploymentBucketName" is 1) the name to call it by in the CFT & 2) overwritten by the logic that identifies that a custom bucket name has been supplied.

I will try and take a look if there's a case there being overlooked in the morning.

erikerikson replied:
The logical names, by which I mean the attribute names under "Resources" in the CFT, are not affected by the custom bucket changes (see [configureStack.js#L98](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/configureStack.js#L98)).  If a custom bucket name is not supplied then it will use the default value defined in the core template (see [core-cloudformation-template.json](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/core-cloudformation-template.json#L12)).

erikerikson commented:
@dougmoscrop, et al.  I have moved the naming object into an attribute of the SDK object.  This let me get rid of the horrible configure idiom and cleaned things up substantially.  Thanks for the great advice!

erikerikson commented:
I have merged the latest changes off of master.

@eahefnawy I feel reasonably confident about these changes since they aren't changing any functionality but only placing hard-coded names and name coercion into a centralized module.  This should, of course, be a no-op.

To verify these changes, I have found that adding a prefix and or suffix for all the names in the following locations can turn up a lot of missed corners:
```
~/lib/plugins/aws/lib/naming.js
~/lib/plugins/aws/deploy/lib/core-cloudformation-template.json
~/lib/plugins/aws/deploy/lib/iam-policy-lambda-execution-template.json
~/lib/plugins/aws/deploy/lib/iam-role-lambda-execution-template.json
```
I am getting 100% success on the unit tests which include the custom bucket naming tests as well.

erikerikson commented:
Debugging through this with live deployment now.  The region and stage are undefined and causing an issue.

erikerikson commented:
Turns out that the provider is not propagating the serverless.yml declared stage and/or region to the serverless.config...  Does it seem more appropriate to transfer those on load or to write contingent use code throughout the aws plugin?  The former seems preferable from a terseness standpoint, with a rule that declares "upon loading, the aws plugin will transfer default values declared in the provider specific manner into the global SLS configuration":
```
    // Transfer Provider supplied default values into the global Serverless configuration hive
    if (this.serverless.service.provider.stage) {
      this.serverless.config.stage = this.serverless.config.stage ||
                                      this.serverless.service.provider.stage;
    }
    if (this.serverless.service.provider.region) {
      this.serverless.config.region = this.serverless.config.region ||
                                      this.serverless.service.provider.region;
    }
```

erikerikson commented:
I successfully deployed using this latest commit but only the hello world default.  I'm operational now in an educational account but don't have anything complicated to real-world test with quite yet.

dougmoscrop commented:
Erik I don't know if my refactor PR would make this easier on you? I mean
over everything to json and reduced a lot of duplicate code. I don't mind
rebasing in your changes but I would be doing them by hand. I guess you
would have to do the same. Were both hittinf similar pain points

On Thu, Oct 6, 2016, 5:01 PM Erik Erikson <notifications@github.com> wrote:

> *@erikerikson* commented on this pull request.
> ------------------------------
>
> In lib/plugins/aws/lib/naming.js
> <https://github.com/serverless/serverless/pull/2257>:
>
> > +    return 'IamPolicyLambdaExecution';
> +  }
> +
> +  getLogGroupName(functionName) {
> +    return `/aws/lambda/${functionName}`;
> +  }
> +
> +  /*
> +   * Lambda Function Naming
> +   */
> +  getNormalizedLambdaName(functionName) {
> +    return this.normalizeName(functionName);
> +  }
> +  extractLambdaNameFromArn(functionArn) {
> +    const splitArn = functionArn.split(':');
> +    // TODO the following two lines assumes default function naming?  Is there a better way?
>
> Sorry, I think I wasn't quite clear. This is from authorizers.js#30
> <https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/apiGateway/lib/authorizers.js#L30>.
> It is extracting a function name from a given ARN that should identify a
> lambda that is to be used as the authorizer for apig. I don't believe that
> should contain the logical name that was used in the CFT to create the
> Lambda. Instead, I believe that if I set a function's custom name to be
> something like a-name-with-dashes that I'll end up with an ARN that is
> something like
> arn:aws:lambda:region:${accountId}:function:a-name-with-dashes.
>
> I would expect this code to return dashes rather than a-name-with-dashes
> which I believe is the expected extraction.
>
> Another thought is that the following would seem to be more a appropriate
> authorizer ARN specification (just adding stage):
> arn:aws:lambda:region:${accountId}:function:a-name-with-dashes:${stage}
> In that case, the stage would be what gets split.
>
> Sorry for being a bit confused and context-poor here, just asking
> questions from my curiosity while trying to make sure and capture all the
> name provision and manipulation logic.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2257>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFA7hSZ7Ymeuqb_pqKXy54sD46Nc2ks5qxWGzgaJpZM4KNNHf>
> .
>

erikerikson commented:
Heyya @dougmoscrop, sorry but I hadn't been aware of your PR.  I'll have to take a look.  Broadly speaking, moving out of static text to objects seems positive to me.

I've definitely been attempting not to boil the ocean here, staying narrow and focussed on the naming concern.

Anyway, I'll take a look at what your changes and give this some thought.

erikerikson commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,231 @@
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
```
@eahefnawy @flomotlik @pmuens - sorry to call out.

`this.serverless.config.stage` appears to not be the natural and expected source of stage and region (referenced elsewhere).  These are elsewhere commonly obtained from the options hive, except that they can come from the provider as well.  Is there a single authoritative source for the region and stage values to use?  Is there a reason that the "right answer" isn't being collected and managed with the serverless.config hive?  Thanks!

eahefnawy replied:
hmmm this `this.serverless.config.stage` is a little weird, I dont remember seeing it before. Maybe it slipped in. But you're right, the general source is `this.options.stage` as seen [here in the aws/validate file](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/lib/validate.js#L12), which is common to all aws plugins.

Maybe it's just a matter of naming, and we shouldn't name it `this.options.stage` since it also takes from the provider, but I'm not sure about keeping it in the `serverless.config` object because stage/region are provider specific, while `serverless.config` is more of serverless/service wide config.

flomotlik replied:
for now `this.options.stage` can be used, but we need to clean this up because its super unintuitive at the moment. Probably a helper function in AWS implementation that reads the current stage from config file, options or other places that it is defined. Similar to the credentials PR @erikerikson recently did

eahefnawy replied:
@erikerikson about the stage/region not being available in the context, maybe in that case we could just pass the stage/region to whatever method needs them? this would make those methods "pure" in a sense. That would be similar to how we're doing it with `sdk.request` for example, you just pass it a stage/region. Do you think this setup would fix this current issue?

as for multiple providers, imo multi provider support in itself is not challenging...what's really challenging is **designing a uniform interface for all providers**, simply because each provider is completely different. I honestly am not completely sure how we're going to accomplish that since lots of other providers are still immature in their interface. But we're keeping our eyes open for any opportunities there 😊

erikerikson replied:
re: "maybe in that case we could just pass the stage/region to whatever method needs them?"
Your proposal is what I was attempting to suggest is problematic.  I could do what you suggest and it **would** fix the **immediate** issue.  It creates a long term problem however.  Doing this constrains what can be used as source material for generating a name (admittedly not an immediate concern).  What about the useful `-foo` switch that will be added to the `deploy` command?  Do we add `foo` to this method signature too?  More plausibly, what about the next provider (beyond AWS) that doesn't have a region and a stage parameter?  Do we add that provider's special CLI configurable values to the signature too?  Over the long term, this approach seems to paint us into a corner.

If I'm missing a big point, I apologize, I'm not sure I quite understand what you mean by 'make those methods "pure" in a sense'.  It sounds as if you're suggesting that the interface be made explicit and thereby that it would be very clear how configuration flows and that would provide a kind of purity.  That said, I don't think that the scope of what will be passed is sufficiently compact or well enough known at this point.

What I am proposing is that centrally SLS will have a scheme for managing the configuration.  It already does this with the `Config` object.  That object could be expanded to provide a richer mechanism of interaction.  The interface of the config object can be made explicit and encapsulated so that data protection occurs and each provider cannot mess with shared state (as others see it) except through well defined interfaces.  Each provider will have effectively unpredictable needs.  As a result, we can centrally maintain the configuration with every provider having an altered version of the centralized configuration available only to them.  I'll get back to this and be concrete.

> as for multiple providers

I am not suggesting that we should provide an abstraction that makes interactions with every provider the same (although similarity would be a little helpful from a cognitive load standpoint).  That leads us toward the least common denominator and none of the deeper value each will have over the others.  Instead, I would suggest we can create a means of separating their perspectives on what the configuration is in isolated contexts that would interact (if ever) through well defined data flows.  Something like:

```
class Config { // the central hive
  constructor(id); // this ID would give you your own version of the configuration
  get(key); // this transparently combines the SLS specified configuration state with the Provider specified configuration state
  set(key, value); // this transparently isolates any value changes by a Provider to the provider
}
```

If it became necessary, you could add more:
```
class Config {
  ...
  getFor(id, key);
  setFor(id, key, value);
}
```
And these would alter the value associated with the given key

```
serverless = new Serverless(); // calls serverless.config.set('foo', 'bar'); and serverless.config.set('my', 'val');
aws = new SDK(serverless);
aws.setFoo('caz') // calls serverless.config.set('foo', 'caz'); // serverless.config now contains global = { foo: 'bar' } and aws = { foo: 'caz' }
serverless.printFoo(); // console.log(this.config.get('foo'); // prints "bar"
aws.printFoo(); // console.log(serverless.config.get('foo'); // prints "caz"
serverless.printMy(); // console.log(this.config.get('my'); // prints "val"
aws.printMy(); // console.log(serverless.config.get('my'); // also prints "val"
```

erikerikson replied:
As for what I meant by multi-provider service, I mean that one could do something like:
```
service: myService

provider:
  stage: dev // applies to all providers
  aws:
    runtime: nodejs4.3
    profile: myProfile
    region: us-east-1
  google:
    profile: myGoogleProfile
    region: googleRegion
functions:
  hello:
    handler: handler.hello // uses my google machine learning resource
resources:
  aws:
    Resources: ...
  google:
    // my machine learning resource
```
I would expect that both provider plugins would be loaded and that the google resources would be deployed using the google provider and the aws resources would deployed by that provider.

eahefnawy replied:
> What about the useful -foo switch that will be added to the deploy command? Do we add foo to this method signature too?

You're right, and this is why we have the options object available for all plugins. What we wanted to do at first is to init this provider class just like other core classes, but we decided to keep all the provider specific logic away from the core.

while I totally agree with you, imo I think we should just pass as args for now since changing that would require some more discussion and internal effort.

Would love to get @flomotlik input though 😊

erikerikson replied:
Definitely the immediate and lowest change choice is definitely passing options into the provider.

For the long term that means spreading the knowledge across the codebase of how to correctly prioritize and select the various configuration sources.

erikerikson replied:
Until the point that centralization occurs, of course.

eahefnawy commented:
@erikerikson I've been testing this PR extensively with the following config trying to cover all supported CF resources/logical ids:

```yml
service: serverless-naming

provider:
  stage: beta
  region: us-west-2
  name: aws
  runtime: nodejs4.3
  apiKeys:
    - keyOne
    - keyTwo

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/create
          method: get
          authorizer: auth
      - s3: naming-bucket
      - schedule: rate(10 minutes)
      - sns: naming-topic
  auth:
    handler: handler.auth
```

The PR is pretty solid for all resources, **however** I can't seem to deploy to a stage passed via the CLI. It seems that (as discussed above), we're **always** taking the stage from the provider and not touching the stage passed as an option. The same probably goes for region as well.

Is there a particular reason why you're storing stage/region in `serverless.config`? It's meant more for a serverless/service level config, rather than provider level config. But if you have to, it needs to be also aware of stages/regions passed as options.

erikerikson commented:
@eahefnawy thank you for taking the time to do this investigation.  I was thinking that it might help us all if there was a set of integration tests that I could (maybe optionally) run (in addition to the unit tests) which would run through these scenarios.  I feel bad that my efforts have spent your and other member's time.
As discussed in the code comment where I talk about centralized config (made shortly ago), the CLI options are not available in the context of the AWS provider plugin (`class SDK`).  I think the discussion around that could probably happen most effectively there.  Without having a complete understanding to the reasoning for the compartmentalization of configuration that I am observing I would end up taking a randomish stab in the dark.

erikerikson commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,231 @@
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
```
should this be `_.capitalize` instead?  (this would lower-case every character but the first)  This would be a breaking change since it could change Resource names.

eahefnawy replied:
I'd vote for `_.upperFirst` to leave room for us and the users to use camelCasing

erikerikson replied:
cool, I'll leave this as is

eahefnawy commented:
@erikerikson aah believe me, reviewing your PRs is the best use of my time! I'm learning a lot 😉

For integration tests, we have the [integration tests suite](https://github.com/serverless/integration-test-suite), but it's sloooooow. I generally like to use the `serverless.yml` template above whenever I'm testing because it's pretty inclusive, but still edge cases has to be well thought through (its an art! 😄  )

erikerikson commented:
Sweet!  Thanks for the integration tests pointer @eahefnawy!  Very glad you're getting value and hope that never changes.

erikerikson commented:
@eahefnawy The CLI options should now be prefered over configuration in Serverless class or provider specified defaults (via https://github.com/serverless/serverless/pull/2257/commits/ae3b9b746deef3cf62a1f9ffcf1708f80f26ad2f).  Moved the options into the SDK as discussed and cleaned up a bit as appropriate.
Also, merged to get up to date.

I have not run the integration tests as I have to run to a family obligation for the weekend.

eahefnawy commented:
@erikerikson that's great! thanks tons! 😊 ... we're currently in the process of improving and moving the integration tests to master, so that should make things easier. I'll be playing around with your recent changes again, and run the integration tests against them once we have them in place.

have fun with your family, and thanks for everything 😊 ... will get back to you with updates 👼

erikerikson commented:
This is merged back to conflict free state but I still have to test the corrupted names configuration (altering the naming class to make sure there's not an accidentally correct name)

pmuens commented on lib/plugins/aws/tests/naming.js:
```diff
@@ -0,0 +1,435 @@
+    it('should capitalize the first letter', () => {
+      expect(sdk.naming.normalizeName('name')).to.equal('Name');
+    });
+    it('should have no effect on caps', () => {
```
Maybe some space here so that all tests follow the same pattern. (This pattern might be found in other tests as well)

pmuens commented on lib/plugins/aws/tests/naming.js:
```diff
@@ -0,0 +1,435 @@
+      serverless.service.service = 'myService';
+      expect(sdk.naming.getStackName()).to.equal(`${serverless.service.service}-${options.stage}`);
+    });
+    it('should handle undefined service name', () => {
```
Should it really handle `undefined` or throw an error instead? (same for the `stage` etc.)

pmuens commented on lib/plugins/aws/tests/naming.js:
```diff
@@ -0,0 +1,435 @@
+      });
+    });
+  });
+  it('#getLogicalRoleName()', () => {
```
Why is there no wrapping in a `describe` block?

pmuens commented on lib/plugins/aws/tests/naming.js:
```diff
@@ -0,0 +1,435 @@
+        ],
+      });
+    });
+    it('handles undefined service name', () => {
```
Should it handle the `undefined` case?

pmuens commented on lib/plugins/create/create.js:
```diff
@@ -85,7 +85,7 @@ class Create {
     // rename the service if the user has provided a path via options
     if (servicePath || serviceName) {
-      const newServiceName = serviceName || servicePath.split(path.sep).pop();
+      const newServiceName = serviceName || servicePath.split(path.sep).pop(); // NAME_HERE
```
What does this mean?

pmuens commented on lib/plugins/package/lib/packageService.js:
```diff
@@ -22,11 +22,11 @@ module.exports = {
 
   getServiceArtifactName() {
-    return `${this.serverless.service.service}.zip`;
+    return `${this.serverless.service.service}.zip`; // NAME_HERE
```
What does this comment mean?

pmuens commented on lib/plugins/package/lib/packageService.js:
```diff
@@ -22,11 +22,11 @@ module.exports = {
 
   getFunctionArtifactName(functionObject) {
-    return `${functionObject.name}.zip`;
+    return `${functionObject.name}.zip`; // NAME_HERE
```
What does this comment mean?

erikerikson replied:
The // NAME_HERE comments indicate a name, possible name, or possible naming convention and should be removed prior to merge.

pmuens commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -0,0 +1,225 @@
+'use strict';
```
Should this file live in `lib`? or should we move it to a `utils` directory as it's kind of a utility functionality...

erikerikson commented:
Thank you for the feedback and sorry @pmuens, I marked this as not ready for review.  I will be focusing very strongly on this PR starting Monday, if not sooner.

pmuens commented:
Thanks for the feedback @erikerikson! No worries.

Just took a look into this as we're pushing it hard so that it can be released next week (as it's very valuable). Thanks for all the hard work and all your contributions! 👑

eahefnawy commented:
@erikerikson thanks for your all effort here 😊 . I've just merged your changes after fixing all the conflicts in #2679 ... So closing this one 🙌
